### DISCUSSION TITLE
Lesson 11 doesn't deploys to a testnet

### DISCUSSION THREAD

**Discussion by NeoRusi** [Cyfrin/foundry-full-course-cu#930] [Q&A] [answered]
I am on lesson 11 about nft's at 8:30:40 when Patrick was deploying the BasicNft.sol to a sepoila etc testnet 
and when i run ```make deploy ARGS="--network sepoila"```
I get this error 
```
Error: 
error sending request for url (http://localhost:8545/): error trying to connect: tcp connect error: Connection refused (os error 61)

Context:
- Error #0: error trying to connect: tcp connect error: Connection refused (os error 61)
- Error #1: tcp connect error: Connection refused (os error 61)
- Error #2: Connection refused (os error 61)
make: *** [deploy] Error 1
```
Previously the anvil localhost was not connected( by I don't understand why it could be a problem if I am deploying it to sepoila etc) ?

**Comment by 0xTagger** [comment]
can you please show your Makefile?

**Reply by NeoRusi** [reply]
```
-include .env

.PHONY: all test clean deploy fund help install snapshot format anvil 

DEFAULT_ANVIL_KEY := 0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

help:
	@echo "Usage:"
	@echo "  make deploy [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""
	@echo ""
	@echo "  make fund [ARGS=...]\n    example: make deploy ARGS=\"--network sepolia\""

all: clean remove install update build

# Clean the repo
clean  :; forge clean

# Remove modules
remove :; rm -rf .gitmodules && rm -rf .git/modules/* && rm -rf lib && touch .gitmodules && git add . && git commit -m "modules"

install :; forge install Cyfrin/foundry-devops@0.0.11 --no-commit && forge install foundry-rs/forge-std@v1.5.3 --no-commit && forge install openzeppelin/openzeppelin-contracts@v4.8.3 --no-commit

# Update Dependencies
update:; forge update

build:; forge build

test :; forge test 

snapshot :; forge snapshot

format :; forge fmt

anvil :; anvil -m 'test test test test test test test test test test test junk' --steps-tracing --block-time 1

NETWORK_ARGS := --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast

ifeq ($(findstring --network sepolia,$(ARGS)),--network sepolia)
	NETWORK_ARGS := --rpc-url $(SEPOLIA_RPC_URL) --private-key $(PRIVATE_KEY) --broadcast --verify --etherscan-api-key $(ETHERSCAN_API_KEY) -vvvv
endif

deploy:
	@forge script script/DeployBasicNft.s.sol:DeployBasicNft $(NETWORK_ARGS)

deployMood:
	@forge script script/DeployMoodNft.s.sol:DeployMoodNft $(NETWORK_ARGS)

mintMoodNft:
	@forge script script/Interactions.s.sol:MintMoodNft $(NETWORK_ARGS)

flipMoodNft:
	@forge script script/Interactions.s.sol:FlipMoodNft $(NETWORK_ARGS)

```

**Reply by theirrationalone** [reply]
```makefile
NETWORK_ARGS := --rpc-url http://localhost:8545 --private-key $(DEFAULT_ANVIL_KEY) --broadcast
```

`$(DEFAULT_ANVIL_KEY)` might expects an environment variable interpolation. Please try to hardcode this.

**Comment by shikhar229169** [comment]
Its a typo in sepolia.
Try this:
```
 make deploy ARGS="--network sepolia"
```

**Reply by NeoRusi** [reply]
Thanks, OMG it was sepolia, where did my 3 years of experience go 😂, previously I used the goerliEth chain
But what about the local foundry testnet how to deploy the BasicNft to Anvil ?

**Reply by shikhar229169** [reply]
Start the anvil chain in one terminal. And run ```make deploy``` in another terminal.

**Reply by NeoRusi** [reply]
I was then running 
```
make deploy ARGS="--network sepolia"
```
after successfully deploying this smart contract to a testnet 
and I get for a some reason this output( it sais that there is no smart contracts deployed)
```
2023-09-28T12:08:04.495367Z ERROR apply:ext: evm::cheatcodes: non-empty stderr args=["bash", "/Users/ruslankuliev/Desktop/foundry-nft's/lib/foundry-devops/src/get_recent_deployment.sh", "BasicNft", "11155111", "/Users/ruslankuliev/Desktop/foundry-nft's//broadcast"] stderr="/Users/ruslankuliev/Desktop/foundry-nft's/lib/foundry-devops/src/get_recent_deployment.sh: line 31: jq: command not found\n/Users/ruslankuliev/Desktop/foundry-nft's/lib/foundry-devops/src/get_recent_deployment.sh: line 33: jq: command not found\n/Users/ruslankuliev/Desktop/foundry-nft's/lib/foundry-devops/src/get_recent_deployment.sh: line 35: ((: i<: syntax error: operand expected (error token is \"<\")\n"
Traces:
  [639098] → new DevOpsTools@0xE901d42C79314a7FaddF2e0b1b0B1D941Fd83678
    └─ ← 3192 bytes of code

  [228375] → new MintBasicNft@0x5b73C5498c1E3b4dbA84de0F1833c4a029d90519
    └─ ← 1030 bytes of code

  [59616] MintBasicNft::run() 
    ├─ [35626] DevOpsTools::e374cdf1(00000000000000000000000000000000000000000000000000000000000000400000000000000000000000000000000000000000000000000000000000aa36a7000000000000000000000000000000000000000000000000000000000000000842617369634e6674000000000000000000000000000000000000000000000000) [delegatecall]
    │   ├─ [0] VM::ffi([pwd]) 
    │   │   └─ ← 0x2f55736572732f7275736c616e6b756c6965762f4465736b746f702f666f756e6472792d6e66742773
    │   ├─ [0] VM::toString(11155111 [1.115e7]) [staticcall]
    │   │   └─ ← 0x000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000083131313535313131000000000000000000000000000000000000000000000000
    │   ├─ [0] VM::ffi([bash, /Users/ruslankuliev/Desktop/foundry-nft's/lib/foundry-devops/src/get_recent_deployment.sh, BasicNft, 11155111, /Users/ruslankuliev/Desktop/foundry-nft's//broadcast]) 
    │   │   └─ ← 0x0000000000000000000000000000000000000000
    │   ├─ [0] console::log(Return Data:) [staticcall]
    │   │   └─ ← ()
    │   ├─ [0] console::log(0x0000000000000000000000000000000000000000) [staticcall]
    │   │   └─ ← ()
    │   └─ ← 0x08c379a0000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000144e6f20636f6e7472616374206465706c6f796564000000000000000000000000
    └─ ← "No contract deployed"



== Logs ==
  Return Data:
  0x0000000000000000000000000000000000000000
Error: 
No contract deployed

make: *** [mint] Error 1
```

**Reply by NeoRusi** [reply]
It doesn't mints for a some reason

**Reply by theirrationalone** [reply]
You haven't deployed your BasicNft contract on sepolia with --broadcast command yet.

**Reply by NeoRusi** [reply]
You mean I should run ```make deploy ARGS="--network sepolia" --broadcas```  or what ?

**Reply by theirrationalone** [reply]
You're already doing that. Sorry for again ignoring details. 

Actually you need to install `jq`

```bash
sudo apt install jq
```

Try again after installing jq.

**Reply by NeoRusi** [reply]
Ii should install it globally in my terminal or in the vs code directory
```
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```
Something goes wrong I have the output from above :(.

**Reply by theirrationalone** [reply]
Try this command

```bash
sudo apt-get install jq
```

Command will work only in ubuntu instance bash terminal.

**Reply by NeoRusi** [reply]
```
sudo: apt-get: command not found
```
:(

**Reply by theirrationalone** [reply]
Please try these last ones: 😅

```bash
sudo apt update
sudo apt upgrade
sudo apt install jq -y
```

If you encounter any error while executing commands above then, Please share your bash screenshots with us.

**Reply by NeoRusi** [reply]
```
The operation couldn’t be completed. Unable to locate a Java Runtime.
Please visit http://www.java.com for information on installing Java.
```
On all 3 commands in the bash terminal it gives this output

**Reply by theirrationalone** [reply]
Please share your terminal screenshots.

**Reply by theirrationalone** [reply]
Let's try to install java first:::...

```bash
sudo apt install openjdk-11-jre
java --version
```
after installing java, Please try last commands again....

**Reply by NeoRusi** [reply]
<img width="1070" alt="Screenshot 2023-09-28 at 22 55 22" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/0f8bab35-2bb2-484a-9447-51593adc5227">

**Reply by NeoRusi** [reply]
......... Even Java doesn't installs 
or I should try from terminal ?

**Reply by theirrationalone** [reply]
You didn't told me that you're using `mac os` 😅

If you don't have `brew` then please install it: to check if brew is installed in your mac try this `brew -v`.

brew installation:
```bash
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

you might also need to install libusb:

libusb installation:
```bash
brew install libusb
```
jq installation:
```brew
brew install jq
```

**Reply by NeoRusi** [reply]
I was recently completing the foundry stable coin lesson there was also an issue with this ```brew``` ....
Maybe Patrick should tell about these dependencies and libraries which should be installed before the lesson it will save much more time

**Reply by NeoRusi** [reply]
So I have installed jq what's next ?)

**Reply by NeoRusi** [reply]
OMG we've done this thanks bro 🥳 🥳 🥳 
<img width="1440" alt="Screenshot 2023-09-28 at 23 19 48" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/4c917e5d-d350-4c03-9bc0-2e130509c151">

**Reply by theirrationalone** [reply]
@NeoRusi ! Yeah it worth to be introduced.

annnndd, Voila we done this 🙌🎉

♥ ;`)

**Comment by theirrationalone** [comment]
> Its a typo in sepolia.
> Try this:
> ```
>  make deploy ARGS="--network sepolia"
> ```

How ignorant we're 😂😂.. 
@shikhar229169 ! Thanks buddy ❤

**Comment by GenevaRoc** [comment]
You didn't spell Sepolia right

**Reply by NeoRusi** [reply]
Thanks but rn there is another error the NFT's are not minting

**Comment by NeoRusi** [comment]
:( 
I can't see the PUG maybe I have messed out with the (IPFS)
<img width="150" alt="Screenshot 2023-09-29 at 10 10 17" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/b7ee9629-a127-41f2-b387-dc07d1d7a323">
here is the tokenURI
```
Ipfs://bafybe1937101r76s7egSoobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/2f1lenane=0-PUG.json"
```

**Reply by theirrationalone** [reply]
Please try to do this:

```bash
Ipfs://bafybe1937101r76s7egSoobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4/2f1lenane=0-PUG.json ❌

Ipfs://bafybe1937101r76s7egSoobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4 ✅
```
>>> Deploy and mint your contract with the correct uri above, Again. 

;`)

**Reply by NeoRusi** [reply]
Again :(
<img width="319" alt="Screenshot 2023-09-30 at 23 02 43" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/594d1cdd-eff8-47b5-9ef7-44439ee3c2ca">


here is the updated tokenURI ```Ipfs://bafybe1937101r76s7egSoobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4```

<img width="1440" alt="Screenshot 2023-09-30 at 23 04 25" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/103575933/baeaeeab-f023-45c3-9c3a-ff6c917bfe91">
And the contract was successfully deployed and minted again maybe I should manually deploy Pug.png to IPFS ?

**Reply by theirrationalone** [reply, marked as answer]
Now i just checked that uri out, Sorry it's a wrong invalid PUG uri. 😅

Please use this one: I brought it from patrick's official repo... and trimmed last filename part...:
```bash
ipfs://bafybeig37ioir76s7mg5oobetncojcm3c3hxasyd4rvid4jqhy4gkaheg4
```

It's working in my machine, You can make a cross check by pasting it in your browser address bar.

**Reply by NeoRusi** [reply]
Thanks let see how the dynamic sag section will go, also maybe you can tell men how to run the anvil node with which command so I can also test it on the anvil chain ?

**Reply by theirrationalone** [reply]
Good Luck for Dynamic/mood nft sec. 🍀🍀🍀

It's so simple if you wanna test it onto anvil chain.

First, You need to clear your metamask's nonce data & activity. Metamask -> settings -> advanced -> scroll to down a bit, It'll appear to you.

Second, Connect your anvil with your metamask and setup an account, On successful connection there will be a huge amount of ETH i.e., 999.99 ETH on your chosen/created account underneath anvil network.

Third, Go back to your project where anvil is running, open a separate terminal and please keep anvil running on that terminal (I know you already familiar with this 😁). Now you can deploy your nft contracts on anvil as well as can mint your nfts.

Fourth, After successful minting, Take the `tokenId` (with which you have minted your nft) and nft address. Go to your metamask again, You would definitely be able to find the `NFTs tab` alongside `Tokens tab` in front of `Activity tab`. Why i'm talking this you already did this 😂. That's it, You know the drill.... You will be good to Go.


Cheers, Feeling Awesome & Appreciated, to work with You,

It was a Great Discussion BTW.

Your Pal! ♥

**Reply by NeoRusi** [reply]
Thanks bro good luck :).
